### ISSUE TITLE
sp > Client Side Pages: Can't copy some webpart content

### ISSUE THREAD

**Issue by mikezimm** [pnp/pnpjs#1587]
@juliemturner , sorry not sure if you are the right person to tag on this one

### Category
- [ ] Enhancement
- [x] Bug
- [x] Question
- [ ] Documentation gap/issue

### Version
"@pnp/sp": "^2.0.7",

### Expected / Desired Behavior / Question
When I copy pages with any webpart to another site, it should work consistantly.

### Observed Behavior
I found when I copy certain webparts from site to site (tested different site collections), some pages would not copy.

These are the webparts that have given issues and what specific conditions caused the crash:
- **Hero Webpart** (works if empty or there are no images in the tiles)
         NOTE:  This is an issue even if the image from the tile comes from the Stock Images, Websearch or Your Organization photos.
- **File Viewer**
- **Picture Viewer**

In all cases, I get the same error message in the console:
**Error message:  400  "Server relative urls must start with SPWeb.ServerRelativeUrl"**

It's like these webparts are storing partial urls that can not be reconciled when copying.
What is odd about URLs is if I have a text webpart with a link to the same site, that webpart transfers just fine but it does make it an absolute url.  That's kind of how I'd expect it to work for the other webparts but it doesn't work.

If there is a way to fix it in the copied object before saving, I could do that... just have no idea where to look since that object seems to go on forever.

### Steps to Reproduce
```typescript
    const page2 = await Web("https://tenant.sharepoint.com/sites/Patterns/").loadClientsidePage("/sites/Patterns/SitePages/" + sourcePageName + ".aspx");

    console.log( 'Source page ' + sourcePageName , page2 );

        try {
            const pageCopy2a = await page2.copy(destWeb, sourcePageName, sourcePageName);
            console.log( 'Succeded pasting page ' + sourcePageName , pageCopy2a );

        } catch (e){
            console.log( 'Failed pasting page ' + sourcePageName  );
        }
    } catch (e) {
            console.log( 'Failed getting page ' + sourcePageName  );
    }
```

**Comment by patrick-rodgers** [comment]
This is going to be one we won't fix. The pages API is non-trivial and every web part is stored differently. Copying pages across sites isn't a use case we are looking to cover currently. We would have to special case each web part and scan every page's content etc.

Would absolutely welcome a PR if someone wants to invest the time here.

**Comment by mikezimm** [comment]
@patrick-rodgers , I did some more testing on the Hero webpart (about 10 different pages with various combos of link and image urls coming from SharePoint, Org Asset Library and "Stock Image") and didn't have issues this time.

However, I did notice some that if my original page had the heading hidden through the UI, the copied page had it expanded.

For some reason, I always have to update the layoutType and re-save.  It's not a big deal but I'm wondering why it does not retain it.

```typescript
page.layoutType = 'NoImage'
awaite page.save();
```

BTW, I remember in the Pnp Ask Me Anything last week, @juliemturner  highlighted reading the docs... I just wanted to say whoever wrote the docs for https://pnp.github.io/pnpjs/sp/clientside-pages/#topicheader was very thorough.  I was having issues with copying banner images and found my answer right there in the docs so +1 for that :)

Now that I think about it, it's possible that my original copy attempts had a banner or thumbnail.

Thanks again, I think this could be closed.
